src/lib.rs

```diff
@@ -0,0 +1,763 @@
+        node_file: Option<String>,
+        sep: u8,
+    ) -> &mut GraphReader {
+        self.node_file_path = node_file;
+        self.edge_file_path = edge_file;
+        self.separate = sep;
+        self
+    }
+
+    /// Call function `build` to generate a instance of `Graph` in parallel.
+    /// In this function use `thread_pool_size` threads for reading edge file and node file.
+    /// You can modify the `thread_pool_size` by method `with_thread(thread_size: u64)`,while
+    /// the param `thread_size` need to be larger than 1.
+    ///
+    /// The basic process in this method as follows:
+    /// 1. Set up thread pool with the configurations provided by user,
+    /// 2. Build a empty instance of `Graph` for filling later,
+    /// 3. Create index for edge file used to separate file for threads,
+    ///      and obtains edge key names from file,
+    /// 4. Construct channel for threads' communication,
+    /// 5. Distribute reading task for threads and run them,
+    /// 6. Collecting thread's result on last thread in pool,
+    /// 7. (Optional) Reading vertices file by parallel if there exist,
+    /// 8. Generating index for vertices vector and edges vector.
+    pub fn build<T>(&mut self) -> Graph<T>
+    where
+        T: Eq + Hash + FromStr + Display + Clone + Copy + Sync + Send + Ord,
+        <T as FromStr>::Err: Debug,
+        for<'de> T: Deserialize<'de>,
+    {
+        self.thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(self.thread_pool_size as usize)
+            .build()
+            .unwrap();
+
+        let mut g = Graph {
+            graph_type: self.graph_type.clone(),
+            vertices: vec![],
+            edges: vec![],
+            vertex_key: vec![],
+            edge_key: vec![],
+            neighbour_index: HashMap::new(),
+            vertex_index: HashMap::new(),
+        };
+
+        let directed = match self.graph_type {
+            GraphType::Directed => true,
+            GraphType::Undirected => false,
+        };
+        let has_node_file = self.node_file_path.is_some();
+
+        let sep = self.separate;
+        let edge_file = &self.edge_file_path;
+        // preserve a thread for collecting
+        let worker_count = self.thread_pool_size - 1;
+
+        // 0.Create index for separating file
+        let mut rdr = ReaderBuilder::new()
+            .has_headers(true)
+            .delimiter(sep)
+            .from_path(edge_file)
+            .expect("Please check `edge_file` path");
+        let mut wtr = Cursor::new(vec![]);
+        RandomAccessSimple::create(&mut rdr, &mut wtr);
+        let mut idx = RandomAccessSimple::open(wtr).unwrap();
+        // Obtains key names from file
+        for key in rdr.headers().unwrap().iter() {
+            g.edge_key.push(key.to_string());
+        }
+
+        // 1. Construct channel for threads' communication,
+        //    Transmit Box pointer here to avoid large memory copy.
+        let (tx, rx) = channel();
```

Comment:
Can you try to access the memory directly, instead of sending via some channels. You know there are multiple ways of  multi-thread communication, pipe and shared-memory are two of them. I think `channel` may fall into the first and is the less efficient one. Consider using shared-memory, using some critical area protection scheme (check `std::sync` package)